### ISSUE TITLE
Reverse proxy code broken

### ISSUE THREAD

**Issue by Adje** [domoticz/domoticz#5199]
Today i found out that de settings for the reverse proxy are gone. 
I have my domoticz install behind a nginx reverse proxy. I whitelisted the ip-range of my home-network so no login is needed. My reverse proxy is in the same ip-range so it didn't need an password either.This was solved by the setting for the "RemoteProxyIP" so everyone outside of my homenetwork needed a password to access by the reverse proxy.

In the  code submit of 9th of march i see "Remove old proxy code".  Probably the setting and with that the security measure for asking the login credentials disappeared. Is this by design and do i need to find a new way to set up my reverse proxy or is this a bug?

**Comment by kiddigital** [comment]
@Adje , can you provide an example flow of what did work and now doesn't work anymore?

**Comment by gizmocuz** [comment]
> "In the code submit of 9th of march i see "Remove old proxy code". Probably the setting and with that the security measure for asking the login credentials disappeared"

This was related to the deprecated MyDomoticz connection and is not related to your problem

**Comment by Adje** [comment]
@kiddigital Not sure how i could provide an example. 

I have in the general settings 192.168.1.* as the network alias for local networks that doesn't need a username and password. 
Before the update i had the option to set the ip of my reverseproxy server. (192.168.1.51) 

If i was outside of my own network en used my url to get to the domoticz gui the login was prompted. In the current version there isn't a login window and the domoticz dashboard is inmediatly opened. The gui is now open for everyone from outside my network. 

My nginx install is configurated like https://www.domoticz.com/wiki/WebServer_Proxy

Note the line at the bottom of that page. 

_For the "X-Forwarded-For" and "X-Real-IP" to be recognized in Domoticz, make sure you enter the IP address of your proxy server (which could be on the same machine like 127.0.0.1) in the main Domoticz setup page underneath "RemoteProxyIPs"_

This setting is not visibel anymore.

**Comment by kiddigital** [comment]
Thank you for the additional info. I think I understand what you are trying to do (or at least did).

Maybe indeed the situation you describe does not work anymore because of the removal of some code

(removed from cWebem.cpp)
```
		if (!myWebem->myRemoteProxyIPs.empty())
			{
				for (auto &myRemoteProxyIP : myWebem->myRemoteProxyIPs)
				{
					if (session.remote_host == myRemoteProxyIP)
					{
						const char *host_header = request::get_req_header(&req, "X-Forwarded-For");
						if (host_header != nullptr)
						{
							if (strstr(host_header, ",") != nullptr)
							{
								//Multiple proxies are used... this is not very common
								host_header = request::get_req_header(&req, "X-Real-IP"); //try our NGINX header
								if (!host_header)
								{
									_log.Log(LOG_ERROR, "Webserver: Multiple proxies are used (Or possible spoofing attempt), ignoring client request (remote address: %s)", session.remote_host.c_str());
									rep = reply::stock_reply(reply::forbidden);
									return;
								}
							}
							session.remote_host = host_header;
						}
					}
				}
			}
```

Although it was originally intended differently I think, it could have been used as you describe.

There are some more security related changes underway and I will have a look to see if/how your scenario could/should be supported.

**Comment by Adje** [comment]
thx for the quick replay. 

If there is a better way to connect a url to the domoticz gui i'm definitley up for that. For now i thought i was following the suggested method in de wiki and it doesn't look that weird. 

From a security point of view i prefer NGINX open to the internet above the server of Domoticz. I use the reverse proxy for multiple services that i host at home.

**Comment by gordonb3** [comment]
> Thank you for the additional info. I think I understand what you are trying to do (or at least did).
> 
> Maybe indeed the situation you describe does not work anymore because of the removal of some code
> 
> (removed from cWebem.cpp)
> 
> ```
> [code]
> ```
> 
> Although it was originally intended differently I think, it could have been used as you describe.
> 
> There are some more security related changes underway and I will have a look to see if/how your scenario could/should be supported.

Nope. This was exactly intended to do what @Adje described and more.

**Comment by gordonb3** [comment]
See pull #2223 and #2246, #2250 for background on this mistakenly deleted bit of code.

**Comment by gizmocuz** [comment]
Yep seems to much was removed as the code from MyDomoticz was used to enable this.

Because we need to release a new stable before we can any of the new code of KidDigital this needs to be resolved first.

proxy wise I don't think it makes any difference if you open a direct port to Domoticz or not. It just proxies. You do have more control when using NGINX

I am going to setup a test environment to test this. I would expect that the headers "X-Forwarded-For" and "X-Real-IP" contain the IP address of the outside user, and this could/should be matched by the local networks setup.
This is how it worked before the MyDomoticz implementation and should work again.
There should be no need for additional whitelists

**Comment by gordonb3** [comment]
The point about the whitelist is that you can spoof these headers, allowing bypass of the login prompt.

**Comment by gizmocuz** [comment]
Yes and that's not what we want for sure... Isn't NGINX setting the forwarded-for/real-ip to the actual IP and overwrites a possible spoofed one ?

**Comment by gordonb3** [comment]
IF you have a proxy machine, yes. If someone who does not access DZ through a proxy though does something like `curl -H "X-Real-IP:127.0.0.1" <switch command>` there is a pretty big chance that this will be executed even though he is supposed to pass credentials. That's what the whitelist was for - to only accept these origin IP changing fields from machines that are supposed to send them.

**Comment by gizmocuz** [comment]
Okey clear.... To bad is was mixed with the MyDomoticz (MyProxy) code... I will try to revert this asap

**Comment by kiddigital** [comment]
I think there is no need for an explicit proxy whitelist. Combine it with the 'trusted' IP list and allow to accept proxy headers when they come from a host in the 'trusted' IP list.

**Comment by gizmocuz** [comment]
Yes thats a better way... I hope it works in the latest beta version, and we indeed could rebuild it to work with the whitelist as two lists is one to much

**Comment by gizmocuz** [comment]
@Adje , is it possible to test if it's working with the latest beta ?

**Comment by Adje** [comment]
@gizmocuz  Just updated to the last beta

All settings are back ✅ (including the IP of the proxyserver)
no login needed from inside the homenetwork using the URL connected to the reverse proxy ✅
Login needed when coming from outside the homenetwork ✅
Made a donation ✅ 

Thank you very much everyone, heel erg bedankt iedereen!

**Comment by gizmocuz** [comment]
Thanks for verifying !

@kiddigital , I was working on removing the extra proxy list and converting this to the 'trusted' IP list, but what this also does is giving admin rights to that machine as well.
Not sure if this is going to be a issue ? But for example you have a NGINX proxy setup on 192.168.20.3 and added this IP address to the local networks.
It will then check if there are forwarded-for headers and use this IP address as new remote IP, this one is not in the trusted IP addresses so this will be OK.
But if there is a user logged on at this machine (192.168.20.3) via SSH, this user has full control of Domoticz.

IF... NGINX is configured wrongly and the proxy headers are not added, this will also make a full open system.
(Sure that this is of course a configuration failure)

It's easy to convert it, so let me know what you think (yes/no)

**Comment by kiddigital** [comment]
Yes, single IP whitelist instead of a separate Proxywhitelist.

Every wrong configuration could lead to an unwanted (open) situation, but that is human error. Can't prevent everything 😁

**Comment by gizmocuz** [comment]
Okey and the fact that it whitelists that whole machine ? (simple SSH user can gain full control)

**Comment by kiddigital** [comment]
In my opinion the whitelist IP range is a list of machines/devices that are allowed to circumvent several security aspects that are normally required.
- bypass authentication
- get Admin privileges
- overwrite host IP addresses (proxy)
- ...

Instead of having separate lists for every possible exception.

When token support gets added, it could become possible to add extra access rights (for example the 'proxy' right) and proxies can pass a token with this right telling Domoticz that it is allowed to send proxy headers that can be used. Then a proxy server does not need to be on the whitelist anymore.

**Comment by gordonb3** [comment]
Unsure how you picture this in your mind, but as it is highly probable that users may want a single machine to hold some of these security exceptions but not all, this does not change the internal logic of the application which should still check for this particular right. In particular since this is about security you should take a lot of care not to break it again and of course if you should change the internal representation (token, bit flag, something other) then any existing rules need to be migrated to the new method.

**Comment by gizmocuz** [comment]
Make a new PR that merges the proxy IP's with the local networks

https://github.com/domoticz/domoticz/pull/5203

**Comment by Adje** [comment]
@kiddigital @gizmocuz i noticed that with the merge the setting proxyip dissapeard again. I added the ip of the reverseproxy to the whitelist. Not that intuitive but the wiki explains it so for me ok.

What i don't understand and not sure if it's a problem is the following: I have now whitelisted 192.168.1.* (all internal devices) and 192.168.1.99 (ip of reverseproxy) Based on the 1.* already listed i would reckon the 1.99 shouldn't be needed. But if i don't at 1.99 all traffic from outside my network is not prompted for the username and password.

I hope made myself clear otherwise i could try it in dutch.

**Comment by gizmocuz** [comment]
You need to add the IP address of your proxy server in the whitelist (192.168.1.99), only then it will be checked if this connection has the proxy/forward headers

**Comment by gizmocuz** [comment]
As a matter of fact, your IP address should have been converted/added automatically to the whitelist by the upgrade. It should not have be needed to do this manually

**Comment by Adje** [comment]
> As a matter of fact, your IP address should have been converted/added automatically to the whitelist by the upgrade. It should not have be needed to do this manually

I had to do this manually. Proxy-IP was not transfert to the whitelist automatically
